### PR TITLE
Question & Answers administration 

### LINKED ISSUE #6: Create questions with answers in an election
Ref.: VOC-ADM-02

As an administrator I want to be able to create new Questions and Answers. These will be a new resource for an election.

The configuration of this models must contemplate:

* it must be possible to have multiple Questions on a given Election
* it must be possible to have different Answers for a given Question
* an Election must have at least a Question
* a Question must have at least two Answers

About the field types:

- Questions have a title and a description
- Answers have a title, a description and a value. Title and values must be unique. The value can be set up by the system automatically with an index (0, 1, etc).

Acceptance criteria: 

- [ ] As an administrator I can create a question inside an election
- [ ] As an administrator I can update a question inside an election
- [ ] As an administrator I can delete a question inside an election
- [ ] As an administrator I can create answers related to a question
- [ ] As an administrator I can update answers related to a question
- [ ] As an administrator I can delete answers related to a question

### PR DESCRIPTION
**Pull request by andreslucena**
This PR is the second part of the `decidim-elections` extraction. 

In this case, we're extracting the administration of the models Questions & Answers. 

Fixes #6 

## Testing

- [ ] As an administrator I can create a question inside an election
- [ ] As an administrator I can update a question inside an election
- [ ] As an administrator I can delete a question inside an election
- [ ] As an administrator I can create answers related to a question
- [ ] As an administrator I can update answers related to a question
- [ ] As an administrator I can delete answers related to a question

### CODE REVIEW

File: lib/decidim/vocdoni/test/factories.rb [hunk 1]
```diff
@@ -62,4 +62,58 @@
       end
     end
   end
+
+  factory :question, class: "Decidim::Vocdoni::Question" do
```
Comment (microstudi) [inline lib/decidim/vocdoni/test/factories.rb:66]:
```suggestion
  factory :vocdoni_question, class: "Decidim::Vocdoni::Question" do
```

File: lib/decidim/vocdoni/test/factories.rb [hunk 2]
```diff
@@ -62,4 +62,58 @@
...
+      answers { 10 }
+    end
+
+    trait :projects do
+      complete
+      answers { 6 }
+    end
+  end
+
+  factory :election_answer, class: "Decidim::Vocdoni::Answer" do
```
Comment (microstudi) [inline lib/decidim/vocdoni/test/factories.rb:97]:
```suggestion
  factory :vocdoni_election_answer, class: "Decidim::Vocdoni::Answer" do
```

File: spec/commands/decidim/vocdoni/admin/create_answer_spec.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+require "spec_helper"
+
+describe Decidim::Vocdoni::Admin::CreateAnswer do
+  subject(:command) { described_class.new(form) }
+
+  let(:organization) { component.organization }
+  let(:participatory_process) { component.participatory_space }
+  let(:component) { election.component }
+  let(:question) { create :question, election: election }
```
Comment (microstudi) [inline spec/commands/decidim/vocdoni/admin/create_answer_spec.rb:11]:
```suggestion
  let(:question) { create :vocdoni_question, election: election }
```

File: spec/commands/decidim/vocdoni/admin/create_answer_spec.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  let(:election) { create :election }
```
Comment (microstudi) [inline spec/commands/decidim/vocdoni/admin/create_answer_spec.rb:12]:
```suggestion
  let(:election) { create :vocdoni_election }
```

File: spec/commands/decidim/vocdoni/admin/create_question_spec.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+require "spec_helper"
+
+describe Decidim::Vocdoni::Admin::CreateQuestion do
+  subject { described_class.new(form) }
+
+  let(:organization) { current_component.organization }
+  let(:participatory_process) { current_component.participatory_space }
+  let(:current_component) { election.component }
+  let(:election) { create :election }
```
Comment (microstudi) [inline spec/commands/decidim/vocdoni/admin/create_question_spec.rb:11]:
```suggestion
  let(:election) { create :vocdoni_election }
```

File: spec/commands/decidim/vocdoni/admin/destroy_answer_spec.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+describe Decidim::Vocdoni::Admin::DestroyAnswer do
+  subject(:command) { described_class.new(answer, user) }
+
+  let(:election) { create :election }
```
Comment (microstudi) [inline spec/commands/decidim/vocdoni/admin/destroy_answer_spec.rb:8]:
```suggestion
  let(:election) { create : vocdoni_election }
```

File: spec/commands/decidim/vocdoni/admin/destroy_answer_spec.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  let(:question) { create :question, election: election }
```
Comment (microstudi) [inline spec/commands/decidim/vocdoni/admin/destroy_answer_spec.rb:9]:
```suggestion
  let(:question) { create :vocdoni_question, election: election }
```

File: spec/commands/decidim/vocdoni/admin/destroy_answer_spec.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  let!(:answer) { create :election_answer, question: question }
```
Comment (microstudi) [inline spec/commands/decidim/vocdoni/admin/destroy_answer_spec.rb:10]:
```suggestion
  let!(:answer) { create :vocdoni_election_answer, question: question }
```

File: spec/commands/decidim/vocdoni/admin/destroy_question_spec.rb [hunk 9]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+describe Decidim::Vocdoni::Admin::DestroyQuestion do
+  subject { described_class.new(question, user) }
+
+  let(:election) { create :election }
+  let!(:question) { create :question, election: election }
```
Comment (microstudi) [inline spec/commands/decidim/vocdoni/admin/destroy_question_spec.rb:9]:
```suggestion
  let(:election) { create :vocdoni_election }
  let!(:question) { create :vocdoni_question, election: election }
```

File: spec/commands/decidim/vocdoni/admin/update_answer_spec.rb [hunk 10]
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+describe Decidim::Vocdoni::Admin::UpdateAnswer do
+  subject(:command) { described_class.new(form, answer) }
+
+  let(:election) { create :election }
+  let(:question) { create :question, election: election }
+  let(:answer) { create :election_answer, question: question }
```
Comment (microstudi) [inline spec/commands/decidim/vocdoni/admin/update_answer_spec.rb:10]:
```suggestion
  let(:election) { create :vocdoni_election }
  let(:question) { create :vocdoni_question, election: election }
  let(:answer) { create :vocdoni_election_answer, question: question }
```

File: spec/commands/decidim/vocdoni/admin/update_question_spec.rb [hunk 11]
```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+describe Decidim::Vocdoni::Admin::UpdateQuestion do
+  subject { described_class.new(form, question) }
+
+  let(:election) { create :election }
+  let(:question) { create :question, election: election }
```
Comment (microstudi) [inline spec/commands/decidim/vocdoni/admin/update_question_spec.rb:9]:
```suggestion
  let(:election) { create :vocdoni_election }
  let(:question) { create :vocdoni_question, election: election }
```

File: spec/forms/decidim/vocdoni/admin/answer_form_spec.rb [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+  let(:context) do
+    {
+      current_organization: component.organization,
+      current_component: component,
+      election: election,
+      question: question
+    }
+  end
+  let(:election) { question.election }
+  let(:question) { create :question }
```
Comment (microstudi) [inline spec/forms/decidim/vocdoni/admin/answer_form_spec.rb:17]:
```suggestion
  let(:question) { create :vocdoni_question }
```

File: spec/forms/decidim/vocdoni/admin/question_form_spec.rb [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+  subject { described_class.from_params(attributes).with_context(context) }
+
+  let(:context) do
+    {
+      current_organization: component.organization,
+      current_component: component,
+      election: election
+    }
+  end
+  let(:election) { create :election }
```
Comment (microstudi) [inline spec/forms/decidim/vocdoni/admin/question_form_spec.rb:15]:
```suggestion
  let(:election) { create :vocdoni_election }
```

File: spec/models/decidim/vocdoni/answer_spec.rb [hunk 14]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+describe Decidim::Vocdoni::Answer do
+  subject(:answer) { build(:election_answer) }
```
Comment (microstudi) [inline spec/models/decidim/vocdoni/answer_spec.rb:6]:
```suggestion
  subject(:answer) { build(:vocdoni_election_answer) }
```

File: spec/models/decidim/vocdoni/question_spec.rb [hunk 15]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+describe Decidim::Vocdoni::Question do
+  subject(:question) { build(:question) }
```
Comment (microstudi) [inline spec/models/decidim/vocdoni/question_spec.rb:6]:
```suggestion
  subject(:question) { build(:vocdoni_question) }
```

**Review by microstudi — COMMENTED** [review]
In this PR, the only thing I'd change would be to namespace the factories. I thing it's a good practice to avoid any confusion with the official election module.

**Comment by andreslucena** [comment]
This is ready to be reviewed again @microstudi 
A few comments: 

* there were a few more factories that we needed to change
* as we're only changing the factory name and not the association name, I also needed to change it in the factories definition at 0a83cd19c222af1a25987fe28193ede97deb13e3
* as the "`publicable`" shared example tries to infer the factory name from the model name, I needed to change that too. There are more details at 96dd7dbae789a8f466b320d0fe2794e947805a9d

File: lib/decidim/vocdoni/test/factories.rb [hunk 16]
```diff
@@ -62,4 +62,58 @@
...
+    end
+
+    association :election, factory: :vocdoni_election
+    title { generate_localized_title }
+    weight { Faker::Number.number(digits: 1) }
+
+    trait :complete do
+      after(:build) do |question, evaluator|
+        overrides = { question: question }
+        question.answers = build_list(:election_answer, evaluator.answers, overrides)
```
Comment (microstudi) [inline lib/decidim/vocdoni/test/factories.rb:78]:
```suggestion
        question.answers = build_list(:vocdoni_election_answer, evaluator.answers, overrides)
```

**Review by microstudi — COMMENTED** [review]
Thanks @andreslucena , amazing job. Looks pretty neat now.

I found a little one more to be rename (I think, see the comment)

**Review by microstudi — APPROVED** [review]
great

**Pull request merged**